charon/temperature-control/temperature-control.ino

```diff
@@ -27,41 +34,65 @@ const int LOOP_DELAY = 500; // ms
 // We want the algae to stay at a certain temperature
 const float setpoint = 10.0; // °C
 
+// variable to store temperature tolerance for implementing hysteresis
+float setpoint_hysteresis = 0.5; // °C
+
+// variable to track duration of peltier on and off-time
+int cooling_duration = 0;
+
+static int uart_putchar (char c, FILE *stream)
+{
+    Serial.write(c) ;
+    return 0 ;
+}
+
 void setup() {
   pinMode(peltierOut, OUTPUT);
   pinMode(sensorPin, INPUT);
+
+  analogReference(INTERNAL); // set ADC to internal bandgap Vref of 1.1V 
+  
   Serial.begin(9600);
+  
+  // set up stdout to uart for printf
+  fdev_setup_stream(&uart, uart_putchar, NULL, _FDEV_SETUP_WRITE);
+  stdout = &uart;
 }
 
-float readingToCelcius(float analogReading) {
-  // Convert reading to V -> 5 volts / 1024 units = 0.0049 volts (4.9 mV) per unit
-  float volts = analogReading * 0.0049;
+float readingToCelsius(float analogReading) {
+  // Convert reading to mV -> 1100 millivolts / 1024 units = 1.074 per unit
+  float millivolts = analogReading * 1.074;
   
-  // The TMP36 reads 0.5 V at 0°C and has an output scale factor of 10 mV/°C.
+  // The TMP36 reads 500 mV at 0°C and has an output scale factor of 10 mV/°C.
   // so [reading in mV - 500 / 10] = [reading in V * 100 - 50] = °C
-  return volts * 100 - 50;
+  return (millivolts - 500) / 10.0;
 }
 
 void loop() {
   // Average many samples
   sum = 0;
   for (int n = 0; n < N_SAMPLES; n++) {
-    sum += readingToCelcius(analogRead(sensorPin));
+    sum += analogRead(sensorPin);
     delay(SAMPLE_FREQUENCY);
+    cooling_duration += SAMPLE_FREQUENCY;
   }
-  temperature = sum / N_SAMPLES;
-
-  Serial.print(temperature);
-  Serial.print("\n");
+  
+  temperature = readingToCelsius((sum / N_SAMPLES));
+  printf("Temp: %+2.2f °C\n", temperature);
 
-  if (temperature > setpoint) {
+  if (temperature > (setpoint + setpoint_hysteresis)) {
     digitalWrite(peltierOut, LOW);
-    Serial.print("turn on the peltiers\n");
+    Serial.print("Peltiers ON: off for %.1f seconds\n");
```

Comment:
Did you want to define `dur_sec` here, too?

Reply:
yep! that also would be why this didn't cause an issue as it needs to be a call to printf as well